zkml/src/layers/provable/mod.rs

```diff
@@ -0,0 +1,502 @@
+mod error;
+
+use anyhow::{Result, anyhow, ensure};
+use ff_ext::ExtensionField;
+use serde::{Deserialize, Serialize, de::DeserializeOwned};
+use std::{
+    collections::{BTreeMap, HashMap},
+    fmt::Debug,
+};
+use transcript::Transcript;
+
+use crate::{
+    commit::precommit::PolyID, iop::{
+        context::{ContextAux, ShapeStep},
+        verifier::Verifier,
+    }, lookup::context::LookupWitnessGen, model::trace::StepData, padding::{PaddingMode, ShapeInfo}, tensor::{ConvData, Number}, Claim, Element, Prover, ScalingFactor, ScalingStrategy, Tensor
+};
+
+use super::{Layer, LayerCtx, LayerProof, flatten::Flatten, requant::Requant};
+
+pub(crate) type NodeId = usize;
+
+pub use error::ProvableOpError;
+
+/// Represents a link between an input/output wire of a node with an input/output wire of
+/// another node.
+#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
+pub struct Edge {
+    // Reference to the node linked to this wire, will be `None` if the wire is an input or
+    // output of the model
+    pub(crate) node: Option<NodeId>,
+    // The index of the wire of `node` which is linked to this wire
+    pub(crate) index: usize,
+}
+
+impl Edge {
+    pub fn new(node: NodeId, index: usize) -> Self {
+        Self {
+            node: Some(node),
+            index,
+        }
+    }
+
+    /// Edge when the node is an input or an output of the model
+    pub fn new_at_edge(index: usize) -> Self {
+        Self { node: None, index }
+    }
+}
+
+/// Represents all the edges that are connected to a node's output wire
+#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
+pub struct OutputWire {
+    // needs to be a vector because the output of a node can be used as input to multiple nodes
+    pub(crate) edges: Vec<Edge>,
+}
+
+/// Represents a node in a model
+#[derive(Clone, Debug)]
+pub struct ProvableNode<N> {
```

Comment:
Since we don't have a simple `Node` i would suggest to rename simply to `Node` ? The provable can be inferred from the library's name i believe 😁

Reply:
Renamed in commit https://github.com/Lagrange-Labs/deep-prove/pull/114/commits/63dede2d959f8d5131d50a44a63384a6eee5138d. I also renamed `ProvaleModel` -> `Model`